### PR TITLE
Apply expression part2

### PR DESCRIPTION
**Pull request by vincentsarago**
This PR apply changes for #523 from #524 comments

- update ImageData.apply_expression to return new ImageData object
- deprecate `asset_expression` (except in MultiBaseReader.statistics() method)
- refactor `point()` method to return value and band_names (helps with expression)

### CODE REVIEW

File: rio_tiler/io/base.py [hunk 1]
```diff
@@ -25,6 +26,13 @@
 from ..utils import get_array_statistics
 
 
+def _AssetExpressionWarning():
+    warnings.warn(
+        "asset_expression is deprecated and will be removed in 4.0. Use pure Expression",
+        DeprecationWarning,
+    )
```
Comment (vincentsarago) [inline rio_tiler/io/base.py:33]: avoids code duplication. will be remove once we publish 4.0

File: rio_tiler/models.py [hunk 2]
```diff
@@ -275,12 +275,21 @@ def apply_color_formula(self, color_formula: Optional[str]):
         for ops in parse_operations(color_formula):
             self.data = scale_dtype(ops(to_math_type(self.data)), numpy.uint8)
 
-    def apply_expression(self, expression: str, bands: Optional[List[str]] = None):
+    def apply_expression(
+        self, expression: str, bands: Optional[List[str]] = None
+    ) -> "ImageData":
         """Apply expression in place."""
```
Comment (vincentsarago) [inline rio_tiler/models.py:279]:
```suggestion
        """Apply expression to the image data."""
```

File: rio_tiler/io/base.py [hunk 3]
```diff
@@ -373,6 +382,9 @@ def merged_statistics(  # type: ignore
             Dict[str, rio_tiler.models.BandStatistics]: bands statistics.
 
         """
+        if asset_expression:
+            _AssetExpressionWarning
```
Comment (geospatial-jeff) [inline rio_tiler/io/base.py:386]:
```suggestion
            _AssetExpressionWarning()
```
Same in other places.

Comment (vincentsarago) [reply rio_tiler/io/base.py:386]: 🤦

File: rio_tiler/io/base.py [hunk 4]
```diff
@@ -194,7 +202,7 @@ def preview(self, **kwargs: Any) -> ImageData:
         ...
 
     @abc.abstractmethod
-    def point(self, lon: float, lat: float, **kwargs: Any) -> List:
+    def point(self, lon: float, lat: float, **kwargs: Any) -> Tuple[List, List[str]]:
```
Comment (geospatial-jeff) [inline rio_tiler/io/base.py:205]: Could this return `ImageData` like the other methods?

[the remark this answers is not in the document]
Comment (geospatial-jeff) [reply rio_tiler/io/base.py:205]: Could this return `ImageData` like the other methods?

[the remark this answers is not in the document]
Comment (vincentsarago) [reply rio_tiler/io/base.py:205]: I'll add `PointData` class in another PR

File: rio_tiler/io/base.py [hunk 5]
```diff
@@ -262,7 +271,7 @@ def _get_asset_url(self, asset: str) -> str:
 
     def parse_expression(self, expression: str) -> Tuple:
         """Parse rio-tiler band math expression."""
-        assets = "|".join([rf"\b{asset}\b" for asset in self.assets])
+        assets = "|".join([rf"\b{asset}\B" for asset in self.assets])
```
[the remark this answers is not in the document]
Comment (vincentsarago) [reply rio_tiler/io/base.py:274]: so far I've got 
```python
_re = re.compile(
    "|".join([rf"\b{re.escape(asset)}(?![^\_b])" for asset in self.assets])
)
```
but it doesn't fully work 
```
print(tuple(set(re.findall(_re, "green_b1*red_b1+red_b1/blue_b1+2.0;red_b1"))))
>>> ('red', 'green', 'blue')

print(tuple(set(re.findall(_re, "green_b10*red_b1+red_b1/blue_b1+2.0;red_b1"))))
>>> ('red', 'green', 'blue')

print(tuple(set(re.findall(_re, "greenish_b1*red_b1+red_b1/blue_b1+2.0;red_b1"))))
>>> ('red', 'blue'). # OK, we don't want to match `greenish` 

print(tuple(set(re.findall(_re, "green*red+red/blue+2.0;red;green"))))
>>>('red'). # NOK, we shouldn't match anything
```

**Pull request merged**